pkg/cache/internal/informers_map.go

```diff
@@ -136,27 +143,42 @@ func (ip *specificInformersMap) Start(stop <-chan struct{}) {
 func (ip *specificInformersMap) HasSyncedFuncs() []cache.InformerSynced {
 	ip.mu.RLock()
 	defer ip.mu.RUnlock()
-	syncedFuncs := make([]cache.InformerSynced, 0, len(ip.informersByGVK))
-	for _, informer := range ip.informersByGVK {
+	syncedFuncs := make([]cache.InformerSynced, 0, len(ip.informers))
+	for _, informer := range ip.informers {
 		syncedFuncs = append(syncedFuncs, informer.Informer.HasSynced)
 	}
 	return syncedFuncs
 }
 
 // Get will create a new Informer and add it to the map of specificInformersMap if none exists.  Returns
 // the Informer from the map.
-func (ip *specificInformersMap) Get(gvk schema.GroupVersionKind, obj runtime.Object) (bool, *MapEntry, error) {
+func (ip *specificInformersMap) Get(gvk schema.GroupVersionKind, obj runtime.Object, opts ...client.ListOption) (bool, *MapEntry, error) {
+	// Set default list options and override
+	listOpts := client.ListOptions{
+		Namespace: ip.namespace,
+	}
+	listOpts.ApplyOptions(opts)
+
+	// If configured using a global namespace, use its cache to avoid duplication
+	if ip.namespace == "" {
```

Comment:
So generally this looks like it will work and I like the overall idea of "use specific cache unless configured to use global one" :)

There are however some things to consider:
* If I read this correctly, using a `source.Kind` with a `Namespace` setting and a non-namespaced `specificInformersMap` will result in getting events from all namespaces, which is not expected
* The current approach means that as soon as any namespace is configured, caches will get established for all namespaces an object was ever requested from which is also unexpected
* When thinking about this approach in the context of label selectors I think just a binary "cache for everything if global cache, else distinct cache for every label selector ever used" is not enough, as it may lead to quite a bit of duplication again. Maybe a list of label selectors for which we want a distinct cache and then have the client internally either use the specific cache or do the label-selecting after receiving the objects from the cache?
* More of a nit, but I think it will be better to not re-use `client.ListOptions` for the `Get` on the cache map, as that forces us to keep their feature set on par


At the end of the day thought, this is the point where I can not give you any further feedback if the approach is feasible or has more issues I didn't consider, we will have to wait for @DirectXMan12 to chime in.

But thanks a lot for pushing this forward, its really great to see some movement here :)